### PR TITLE
Changed queueTask callback to an anonymous async function for a 3X performance boost

### LINKED ISSUE #54612: createBufferedTransformStream is unnecessarily slow
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
➜  next.js git:(canary) next info            

    Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.5.0: Thu Jun  8 22:22:23 PDT 2023; root:xnu-8796.121.3~7/RELEASE_ARM64_T6020
    Binaries:
      Node: 20.5.1
      npm: 9.8.0
      Yarn: N/A
      pnpm: 8.6.11
    Relevant Packages:
      next: 13.4.20-canary.9
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
      typescript: N/A
    Next.js Config:
      output: N/A
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

Data fetching (gS(S)P, getInitialProps)

### Link to the code that reproduces this issue or a replay of the bug

https://github.com/vercel/next.js/blob/2b497413087c56dedb3e04f77ea0deebb5d12d0b/packages/next/src/server/stream-utils/node-web-streams-helper.ts#L114

### To Reproduce

git clone https://github.com/wirelessben/Nextjs-async-test.git
➜  Nextjs-async-test git:(main) ✗ node setTimeout.js  
setTimeout: 1.364ms
➜  Nextjs-async-test git:(main) ✗ node setImmediate.js
setImmediate: 0.845ms
➜  Nextjs-async-test git:(main) ✗ node just_async.js  
just async: 0.227ms

### Describe the Bug

An asynchronous callback is ~6X faster than the setTimeout function.

### Expected Behavior

The asynchronous callback will perform much faster than the setTimeout function.

### Which browser are you using? (if relevant)

Chrome Version 116.0.5845.96

### How are you deploying your application? (if relevant)

Command line test app

### PR DESCRIPTION
**Pull request by wirelessben**
### What?
This PR speeds up the createBufferedTransformStream function by about 3X.

### Why?
This change is for performance reasons.

### How?
This change uses an [anonymous async function](https://www.geeksforgeeks.org/javascript-anonymous-functions/) instead of the setTimeout callback.

### Tests

This repo tests the setTimeout callback, setImmediate callback, and just a simple anonymous async function: https://github.com/wirelessben/nextjs-async-test. These are the results:

➜  Nextjs-async-test git:(main) ✗  node setTimeout.js
setTimeout: 1.364ms
➜  Nextjs-async-test git:(main) ✗  node setImmediate.js
setImmediate: 0.845ms
➜  Nextjs-async-test git:(main) ✗  node just_async.js
just async: 0.227ms
➜  Nextjs-async-test git:(main) ✗ node queueTask.js
queueTask: 1.675ms
➜  Nextjs-async-test git:(main) ✗ node queueTask.js
queueTask: 1.538ms
➜  Nextjs-async-test git:(main) ✗ node queueTask.js
queueTask: 0.613ms
➜  Nextjs-async-test git:(main) ✗ node queueTask.js
queueTask: 0.633ms

Fixes #54612

### CODE REVIEW

**Comment by wirelessben** [comment]
I noticed an extraneous semicolon on line 108. Lint-staged won't let me commit an empty commit.  ???

File: packages/next/src/server/stream-utils/node-web-streams-helper.ts [hunk 1]
```diff
@@ -105,12 +105,12 @@ export function createBufferedTransformStream(): TransformStream<
   const flushBuffer = (controller: TransformStreamDefaultController) => {
     if (!pendingFlush) {
       pendingFlush = new Promise((resolve) => {
-        setTimeout(() => {
+        (async () => {
```
Comment (devjiwonchoi) [inline packages/next/src/server/stream-utils/node-web-streams-helper.ts:108]: Missing a semicolon in IIFE `;(async () => // ...`

I'm assuming that the use of `setTimeout` is intentional so as to create a buffer on the next event loop?

Comment (wirelessben) [reply packages/next/src/server/stream-utils/node-web-streams-helper.ts:108]: The async function is properly encapsulated inside another function, so there's no risk of breakage if some tool inlines this file with other files. And five years ago StackOverflow had this:
>I answered this originally in 2013 and I should point out now that most modern javascript web tools are smart enough to handle this now, and with ES2015 and newer, there's better encapsulation, leading to less IIFEs. I haven't had to use an IIFE in years. – 
[Erty Seidohl](https://stackoverflow.com/users/374601/erty-seidohl)
 [Apr 17, 2018 at 20:03](https://stackoverflow.com/questions/17978883/what-is-the-purpose-of-a-semicolon-before-an-iife#comment86789191_17978926)
 
The buffer requirement is satisfied with the first let statement. flushBuffer is just trying to do its job of flushing the buffer, but setTimeout is the worst way to do this, as the callback is put on the bottom of the stack. SetImmediate would be better, as it goes on the top of the stack, but the anonymous async function is clearly the winner, as the tests show.

**Comment by wirelessben** [comment]
Max Leiter found the exact same problem with this function today. His fix was accepted, but this PR is still 3X faster.

The Tests section was updated with the new times.

**Comment by ijjk** [comment]
Allow CI Workflow Run

- [ ] approve CI run for commit: cf28b22aad24effcd5558a6971acc77bf03d80d2

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

**Comment by ijjk** [comment]
Allow CI Workflow Run

- [ ] approve CI run for commit: cf28b22aad24effcd5558a6971acc77bf03d80d2

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

**Comment by ijjk** [comment]
## Failing test suites 
Commit: cf28b22aad24effcd5558a6971acc77bf03d80d2

`pnpm test-dev test/development/acceptance-app/dynamic-error.test.ts` 
- dynamic = "error" in devmode > should show error overlay when dynamic is forced

<details>
<summary>Expand output</summary>

  ● dynamic = "error" in devmode › should show error overlay when dynamic is forced

    expect(received).toMatchInlineSnapshot(snapshot)

    Snapshot name: `dynamic = "error" in devmode should show error overlay when dynamic is forced 1`

    Snapshot: "Error: Page with `dynamic = \"error\"` couldn't be rendered statically because it used `cookies`. See more info here: https://nextjs.org/docs/app/building-your-application/rendering/static-and-dynamic#dynamic-rendering"
    Received: "Error: Connection closed."

      33 |     await session.hasRedbox(true)
      34 |     console.log(await session.getRedboxDescription())
    > 35 |     expect(await session.getRedboxDescription()).toMatchInlineSnapshot(
         |                                                  ^
      36 |       `"Error: Page with \`dynamic = \\"error\\"\` couldn't be rendered statically because it used \`cookies\`. See more info here: https://nextjs.org/docs/app/building-your-application/rendering/static-and-dynamic#dynamic-rendering"`
      37 |     )
      38 |

      at Object.<anonymous> (development/acceptance-app/dynamic-error.test.ts:35:50)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by wirelessben** [comment]
A good suite of tests is worth a thousand opinions. Thanks for the look.

**Pull request closed without merging**